### ISSUE TITLE
Please add support for Metal as GPU accelerator

### ISSUE THREAD

**Issue by robinhad** [numba/numba#5706]
## Support Metal as GPU accelerator

It would be great to add support for Metal as GPU accelerator for Numba to speed up Python on macOS machines.

**Comment by gmarkall** [comment]
Thanks for the request - adding a new target to Numba is quite a lot of work for the initial implementation, and there is a continued maintenance workload for each target. I'm not a core maintainer, but I find it hard to foresee this being added in the near-term without significant extra development effort joining the project.

Do you have some use cases in mind that would really benefit from a Metal target?

**Comment by robinhad** [comment]
@gmarkall, thanks for your reply!
I develop mainly on macOS, and project, I'am currently involved in and developed for macOS/Linux, would benefit greatly if there is a possibility to speed up code using iGPU/eGPU.
I was able to speed up my code using GPU on Linux and want to achieve the same on mac.

**Comment by seibert** [comment]
I did some casual research on this, and was unable to figure out how to target Metal using LLVM directly.  The Apple docs did not make it very clear how to use Metal in a JIT fashion the same way as OpenCL or NVVM.  The docs may have improved since I checked, so if someone can find an example of another project doing this, it would definitely help.

**Comment by robinhad** [comment]
@seibert, maybe this could help?
https://github.com/gzorin/LLAIR

**Comment by AzizAlqasem** [comment]
Adding support for Metal will significantly help a broad class of Mac and Numba users. I am Ph.D. student specialised in physics. Having Numba GPU accelerator on my Mac (m1 pro) will help me parallize “Time Dependent Schrodinger Equation”!  
I wish I can add this feature myself, but this is way beyond my knowledge and experiance.

**Comment by seibert** [comment]
I would still love to see this, but 18 months later, there still is not a good option (that I've found in my semi-yearly searching).  Apple does not seem to be interested in anyone being able to target their GPUs except by writing Metal C++, so they don't expose the LLVM API directly.  (The above LLAIR project still requires you generate code in the C++-based Metal Shading Language.)  Numba doesn't have any support for emitting C++-style code from a backend (rather than LLVM instructions).

**Comment by thipokKub** [comment]
Hello, I am sure I'm not qualified to write metal shader but in some documentation of Metal API it seems like you can create `MTLLibrary` object directly in swift at runtime using [makeLibrary](https://developer.apple.com/documentation/metal/mtldevice/1433431-makelibrary) function. I'm not sure if this is possible but technically, you can connect a compiled metal shader with python through swift (see [noppoMan/python-metal-benchmark](https://github.com/noppoMan/python-metal-benchmark)]. So maybe if the `makeLibrary` function is invoke through swift from python, it can be compiled on the fly?

**Comment by seibert** [comment]
Yeah, that part makes sense.  The hard part is translating Numba's generated code, which comes out of the compiler pipeline in some variant of LLVM IR (very low level), into the Metal Shader Language, which is C++-based (much higher level).  See the [Numba architecture docs](https://numba.readthedocs.io/en/stable/developer/architecture.html?highlight=IR#stage-6a-generate-nopython-llvm-ir) for examples of how this looks.

There have been LLVM IR to C backend projects in the past, but this would be a quite a research project.  The frustrating part is that Apple internally does have an LLVM IR-based backend for Metal, but they don't document it anywhere and reverse engineering it would be a different research project.

**Comment by thipokKub** [comment]
Okay, please correct me if I'm wrong

So the problem is there is no known way to convert numba's LLVM IR code to metal GPU shader target. So I think it is right now next to impossible for anyone to compile an efficient code for metal GPU (except Apple)

[This blog](https://worthdoingbadly.com/metalbitcode/)'s result suggest that for the most part metal bitcode is very similar to the compiled LLVM IR code. Maybe this is just a pipe dream, but maybe it might be possible to construct a compiled unoptimized metal bitcode from LLVM IR by adding a few header, and valid format

**Comment by seibert** [comment]
Yes, this is correct, and this is why I described this as a "research project".  :)  If someone figures this out, that would be a big step.

**Comment by Crear12** [comment]
I just want to propose another use case. I'm a PhD student dealing with point cloud data. Previously, I designed some python scripts with number cuda.jit to process the data in parallel, which reduced the process time from one hour (xeon) to 40 seconds (Quadro P4000). However, it was always painful when the GPU memory reached cap. We had several different solutions, including cut the data, got three more Quadro P4000 for STI. Eventually, with so many efforts and funding, we could only have 4*8GB CUDA memory. But now, the M1 Max has a total of 64GB memory(for both CPU and GPU though) with only 140W power consumption. The support for Metal as GPU accelerator can help us migrate our codes to the Apple Silicon platform smoothly and we can develop new algorithms that can still function on the traditional Nvidia computers as long as we switch the "@metal.jit" back to "@cuda.jit".
I also considered migrating everything onto Tensorflow because Apple says that Tensorflow is officially supported, but our algorithms are not machine learning and most of the parts could not benefit from the accelerated operations provided by Tensorflow.

**Comment by thipokKub** [comment]
Will this work?

1. Convert Python -> `numba` -> `LLVM-IR`
2. Convert `LLVM-IR` -> `SPIR-V` ([KhronosGroup/SPIRV-LLVM-Translator](https://github.com/KhronosGroup/SPIRV-LLVM-Translator))
3. Run `SPIR-V` on vulkan backend [KhronosGroup/SPIRV-Tools](https://github.com/KhronosGroup/SPIRV-Tools)
4. Convert vulkan backend API to metal API [KhronosGroup/MoltenVK](https://github.com/KhronosGroup/MoltenVK)

All of `KhronosGroup` projects are available on homebrew
- **SPIRV-LLVM-Translator** `brew install spirv-llvm-translator`
- **SPIRV-Tools** `brew install spirv-tools`
- **MoltenVK** `brew install molten-vk`

**Comment by Crear12** [comment]
Pytorch now supports Metal GPU on M1 Mac, with the help of the Apple team. Please see if you can get assistance from the macOS&Silicon designers. I really want to see how this feature will benefit from the unified memory. The copy_to_host and to_device functions shall be significantly speeded up.

**Comment by seibert** [comment]
We don't have any special contacts with the Apple developer teams, but if someone out there does, please let us know.  :)

I should note that PyTorch is using Metal the way Apple documents, with ahead-of-time compiled C++ kernels.  As mentioned above, Numba needs a just-in-time compiler interface based on LLVM IR in order for a target to work.  As of now, Apple has not announced such a thing.  (WWDC is in a little over two weeks, so we can always hope, of course.)

**Comment by Crear12** [comment]
I'm not sure if this helps, but someone says that the Metal api itself is actually LLVM: https://github.com/gfx-rs/javelin/issues/4

It sounds like: We don't compile on the device. Don't know the details. "It just works"

**Comment by Crear12** [comment]
Hi It's me again, use this to submit feature request:https://www.apple.com/feedback/macos.html

**Comment by seibert** [comment]
I think our best bet is figuring out how Julia did this: https://github.com/JuliaGPU/Metal.jl
(I'm still not sure how it works, since it seems to delegate most of the work to other parts of Julia's GPU compiler.)

**Comment by giucesar** [comment]
I also would love to have this feature, but I completely understand that it requires time and resources to add it.

Just as a reference, taichi-lang (https://www.taichi-lang.org/) was able to have CUDA, Vulkan, and Metal support at compilation time. In the current project I am working on, I use Numba for the majority of the code, but I am using Taichi for their Autodiff and GPU processing in critical steps (they are far from mature for the entire project). It is a bit annoying to deal with two packages with a big overlap, but that is an alternative while the folks from Numba focus (correctly) on where there will be more users and higher impact.

**Comment by realqhc** [comment]
I think this fork from Julia https://github.com/JuliaGPU/llvm-metal contains a Metal shader backend, I can't test it so not sure if it works.

**Comment by Crear12** [comment]
Hey,
Just want to bring some fresh air here since there hasn't been any update for a while. The following is from ChatGPT:
Me: How is MPS performance from Apple? Does numba support MPS acceleration?
ChatGPT: MPS (Metal Performance Shaders) is a set of low-level APIs for performing computations on Apple devices that use Metal, the company's GPU (graphics processing unit) framework. MPS can be used to accelerate a wide range of computations, including image and video processing, machine learning, and data compression. According to Apple, MPS can provide significant performance improvements compared to using the CPU alone, especially on devices with powerful GPUs.

Numba is an open-source just-in-time (JIT) compiler for numerical Python code. It can be used to accelerate Python code running on a wide range of platforms, including macOS. Numba supports a variety of backends for code execution, including CPU, CUDA, and OpenCL. However, Numba does not currently have built-in support for MPS acceleration.

Therefore, if you wish to use MPS acceleration with Numba you would have to create a custom CUDA kernel using the MPS API and invoke it using Numba CUDA.

Me: create a custom CUDA kernel using the MPS API and invoke it using Numba CUDA
ChatGPT:
Creating a custom CUDA kernel using the MPS API and invoking it using Numba CUDA would involve the following general steps:

Create a new Metal compute function that performs the computation you want to accelerate using MPS.
Use the MPS API to create an MPSComputeCommandEncoder, which is used to encode the Metal compute function and other GPU commands into a command buffer.
Use the MPS API to configure the input and output textures for the compute function, and to set any other parameters or resources that the function needs.
Use the Numba CUDA API to create a CUDA kernel that invokes the Metal compute function.
Use Numba CUDA to allocate GPU memory and transfer data between the CPU and GPU.
Use the Numba CUDA kernel to launch the computation on the GPU.
Use the Numba CUDA API to retrieve the results from the GPU and transfer them back to the CPU.
It's important to note that MPS is only available for Apple devices, so this would not work on other platforms or devices. Also, implementing this solution would require a deep understanding of both the MPS and Numba CUDA APIs and a good understanding of GPU programming.

**Comment by gmarkall** [comment]
> Just want to bring some fresh air here since there hasn't been any update for a while.

Pasting nonsense from ChatGPT simply wastes the time of maintainers and anyone else who takes an interest in this issue - it doesn't bring fresh air. Please do not add ChatGPT transcripts to issues in open-source projects.

**Comment by harrymyburgh** [comment]
Hi there, 

Just checking in whether there is any development being done on this topic? It would be a great help to me if this is implemented!

Regards.

**Comment by gmarkall** [comment]
There's none that I'm aware of. It would be a significant amount of work so I don't expect this will be likely to change unless someone shows up to do or support a large part of it.

**Comment by hg0428** [comment]
This would be great to have!

**Comment by KDcode868** [comment]
Any progress 

I want cuda to metal translation in vm on MacOS

**Comment by Crear12** [comment]
Not sure if you have seen this, Apple releases this so called Metal Shader Converter with macOS 14 Ventura: https://developer.apple.com/metal/shader-converter/, the description says "Metal shader converter converts shader intermediate representations in LLVM IR bytecode into bytecode suitable to be loaded into Metal. It’s available as a library and a standalone executable. All the functionality exposed through the library interface is available via the standalone executable." From what I read from this thread, @thipokKub in 2021 mentioned that "there is no known way to convert numba's LLVM IR code to metal GPU shader target". Can this new Metal Shader Converter make some changes?

**Comment by gmarkall** [comment]
It appears that it converts DXIL, which is a subset of LLVM 3.7 IR. I would guess that this exists in support of the Game Porting Toolkit, so I'm not optimistic about it making a straightforward conversion for the IR that Numba generates. If you want to experiment and post some results converting Numba-generated IR using the Shader Converter, then that would be illuminating though.

**Comment by seanlaw** [comment]
In case it matters, I just discovered that the Metal Performance Shaders (MPS) framework only supports single-precision numbers. Sadly, for my use case, this is a deal breaker 😢

**Comment by John15321** [comment]
Hi, wanted to ask what the progress on Apple Silicon GPU support is for numba?

**Comment by honeyspoon** [comment]
maybe this project could be used in the background
https://github.com/MEHDI342/CUDAM

**Comment by gmarkall** [comment]
> maybe this project could be used in the background https://github.com/MEHDI342/CUDAM

That project appears to parse CUDA C++ code, which is not part of the Numba CUDA pipeline - Numba turns Python into LLVM IR then PTX.

**Comment by eviking** [comment]
Started working on this without knowing about this thread, so I figured I would share my work if anybody can use for anything. Absolutly not complete, but good enough for what I needed it for. [https://github.com/eviking/numba-metal](https://github.com/eviking/numba-metal)
